msal4j-sdk/src/integrationtest/java/com.microsoft.aad.msal4j/OnBehalfOfIT.java

```diff
@@ -121,13 +117,60 @@ void acquireTokenWithOBO_testCache(String environment) throws Exception {
                                 .build()).
                         get();
 
-        assertNotNull(result6);
-        assertNotNull(result6.accessToken());
+        assertResultNotNull(result6);
         assertNotEquals(result6.accessToken(), result5.accessToken());
         assertNotEquals(result6.accessToken(), result4.accessToken());
         assertNotEquals(result6.accessToken(), result2.accessToken());
     }
 
+    @Test
+    void acquireTokenWithOBO_TenantOverride() throws Exception {
+        cfg = new Config(AzureEnvironment.AZURE);
+        String accessToken = this.getAccessToken();
+
+        final String clientId = cfg.appProvider.getOboAppId();
+        final String password = cfg.appProvider.getOboAppPassword();
+
+        ConfidentialClientApplication cca =
+                ConfidentialClientApplication.builder(clientId, ClientCredentialFactory.createFromSecret(password)).
+                        authority(cfg.tenantSpecificAuthority()).
+                        build();
+
+        //This token should be cached with the tenant-specific authority set at the application level
+        IAuthenticationResult resultNoOverride = cca.acquireToken(OnBehalfOfParameters.builder(
+                                Collections.singleton(cfg.graphDefaultScope()),
+                                new UserAssertion(accessToken)).build()).
+                        get();
+
+        //This token should be cached with an 'organizations' authority set at the request level
+        IAuthenticationResult resultOrganizations = cca.acquireToken(OnBehalfOfParameters.builder(
+                                Collections.singleton(cfg.graphDefaultScope()),
+                                new UserAssertion(accessToken))
+                                .tenant("organizations")
```

Comment:
I am not sure I agree with this test. Afaik MSAL will use the tenant ID from the response, not from the request, so it's not that clear what this test does. It's also showcases 2 anti-patterns:

1. go from a tenant specific authority on the CCA to the `lmo/organizations` authority on the request. Usually it's the other way around.
1. OBO with "/common" or "/organizations". This is wrong, OBO should always be tenanted. It should use the client token's `tid` claim. Otherwise you introduce a bug where guest users can't login. See https://learn.microsoft.com/en-us/entra/msal/dotnet/acquiring-tokens/web-apps-apis/on-behalf-of-flow#important-note-on-on-behalf-of-obo-flow-with-guest-users


My recommendation is to rely on unit tests for this one: 

- app authority is { "lmo/organizations", "lmo/T1", "lmo/T2" }
- req tenant is {T1, T2}
- flow { OBO, client_creds, silent}

So there are quite a few unit tests (3x2x3), but mostly repetitive.

Then Assert on the correct MSAL behavior i.e. token from cache or from the correctly tenanted token endpoint

Reply:
Agreed that unit tests would be better, originally I just did an integration test since there were ones to copy from but we didn't have any OBO unit tests.

In the latest commit I replaced this integration test with a couple of unit tests. It's not as many as you suggest, but they should show the correct behavior by checking for the correct number of HTTPClient calls, the token cache size, etc., depending on whether the tenant was set at the request level or not.